File: stapp/app/src/main/java/com/xvesa/stapp/NavigationHandler.java
package com.xvesa.stapp;

import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.core.util.Pair;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import java.util.HashMap;
import java.util.Objects;

public class NavigationHandler {

    public static class AppTabFragment extends Fragment {

        private final int id;

        public AppTabFragment(int id) {
            this.id = id;
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            // Inflate the layout for this fragment
            return inflater.inflate(this.id, container, false);
        }
    }

    public static class HomePageFragment extends AppTabFragment {

        private final HashMap<Integer, AppTabFragment> mapper;

        private Integer lastViewID;

        public HomePageFragment(int id) {
            super(id);
            this.lastViewID = R.id.home_tabs_menu_first;

            this.mapper = new HashMap<>();
            this.mapper.put(R.id.home_tabs_menu_first, new AppTabFragment(R.layout.experiences_layout));
            this.mapper.put(R.id.home_tabs_menu_second, new AppTabFragment(R.layout.adventures_layout));
            this.mapper.put(R.id.home_tabs_menu_third, new AppTabFragment(R.layout.activities_layout));
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return super.onCreateView(inflater, container, savedInstanceState);
        }

        public void afterCreateHandler() {
            this.getChildFragmentManager()
                    .beginTransaction()
                    .add(R.id.main_frame, Objects.requireNonNull(this.mapper.get(R.id.home_tabs_menu_first)))
                    .commitNow();

            RelativeLayout frame = (RelativeLayout) this.getView();
            assert frame != null;
            this.mapper.forEach((id, fragment) -> {
                TextView home_tab = frame.findViewById(id);
                home_tab.setOnClickListener(a -> {
                    TextView last_tab = frame.findViewById(this.lastViewID);
                    last_tab.setTypeface(null, Typeface.NORMAL);
                    last_tab.setTextColor(ContextCompat.getColor(frame.getContext(), R.color.home_tabs_original));

                    this.lastViewID = id;
                    TextView this_tab = frame.findViewById(id);
                    this_tab.setTypeface(null, Typeface.BOLD);
                    this_tab.setTextColor(ContextCompat.getColor(frame.getContext(), R.color.home_tabs_selected));

                    this.getChildFragmentManager()
                            .beginTransaction()
                            .replace(R.id.main_frame, Objects.requireNonNull(this.mapper.get(id)))
                            .commitNow();
                });
            });
        }

    }

    private final HashMap<Integer, AppTabFragment> mapper;
    private final HashMap<Integer, Pair<Integer, Integer>> res_map;
    private final AppCompatActivity activity;

    private Integer currentID;

    public NavigationHandler(AppCompatActivity a) {
        this.activity = a;

        this.mapper = new HashMap<>();
        this.mapper.put(R.id.navigation_home, new HomePageFragment(R.layout.home_layout));
        this.mapper.put(R.id.navigation_discover, new AppTabFragment(R.layout.discover_layout));
        this.mapper.put(R.id.navigation_dashboard, new AppTabFragment(R.layout.dashboard_layout));

        this.res_map = new HashMap<>();
        this.res_map.put(R.id.navigation_home, new Pair<>(R.drawable.home_original, R.drawable.home_clicked));
        this.res_map.put(R.id.navigation_discover, new Pair<>(R.drawable.search_original, R.drawable.search_square));
        this.res_map.put(R.id.navigation_dashboard, new Pair<>(R.drawable.dashboard_original, R.drawable.dashboard_clicked));

        this.currentID = R.id.home_tabs_menu;
    }

    private void handleNavigation() {
        TextView default_tab = this.activity.findViewById(R.id.navigation_home);
        Drawable default_drawable_clicked = ContextCompat.getDrawable(this.activity, R.drawable.home_clicked);
        default_tab.setCompoundDrawablesWithIntrinsicBounds(null, default_drawable_clicked, null, null);
        this.currentID = R.id.navigation_home;

        HomePageFragment default_fragment = (HomePageFragment) Objects.requireNonNull(this.mapper.get(R.id.navigation_home));
        FragmentManager supportFragmentManager = this.activity.getSupportFragmentManager();
        supportFragmentManager
                .beginTransaction()
                .add(R.id.fragment_container, default_fragment)
                .commitNow();

        this.mapper.forEach((key, value) -> {
            TextView bottom_nav_view = this.activity.findViewById(key);
            bottom_nav_view.setOnClickListener(a -> {
                if (this.mapper.containsKey(this.currentID)) {
                    TextView last = this.activity.findViewById(this.currentID);
                    Pair<Integer, Integer> pair = this.res_map.get(this.currentID);
                    assert pair != null;
                    Integer view_original = pair.first;
                    Drawable drawable_original = ContextCompat.getDrawable(this.activity, view_original);
                    last.setCompoundDrawablesWithIntrinsicBounds(null, drawable_original, null, null);
                }

                this.currentID = key;
                TextView here = this.activity.findViewById(key);
                Pair<Integer, Integer> pair = this.res_map.get(key);
                assert pair != null;
                Integer view_clicked = pair.second;
                Drawable drawable_clicked = ContextCompat.getDrawable(this.activity, view_clicked);
                here.setCompoundDrawablesWithIntrinsicBounds(null, drawable_clicked, null, null);

                supportFragmentManager
                        .beginTransaction()
                        .replace(R.id.fragment_container, value)
                        .commitNow();
            });
        });
    }

    public void afterCreateHandle() {
        HomePageFragment default_fragment = (HomePageFragment) Objects.requireNonNull(this.mapper.get(R.id.navigation_home));
        assert default_fragment.getView() != null;
        default_fragment.afterCreateHandler();
    }

    public void onCreateHandle() {
        this.handleNavigation();
    }
}

File: stapp/app/src/main/java/com/xvesa/stapp/NavigationHandler.kt
package com.xvesa.stapp

import android.graphics.Typeface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.HorizontalScrollView
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView


class NavigationHandler(private val activity: AppCompatActivity) {

    /**
     * Base class for DIY Fragments
     * @constructor Passed the id so that it makes the specific layout to be loaded.
     */
    open class AppTabFragment(private val id: Int) : Fragment() {

        /**
         * @return Fragment that is loaded for specified layout.
         */
        override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
        ): View? = inflater.inflate(id, container, false)
    }

    /**
     * Inherited from AppTabFragment
     * @constructor Inherited from the super class and perform the same.
     */
    class HomePageFragment(id: Int) : AppTabFragment(id) {

        /**
         * mapper: Store the home tabs and their respective DIY Fragments.
         */
        private val mapper = mapOf(
            R.id.home_tabs_menu_first to AppTabFragment(R.layout.experiences_layout),
            R.id.home_tabs_menu_second to AppTabFragment(R.layout.adventures_layout),
            R.id.home_tabs_menu_third to AppTabFragment(R.layout.activities_layout)
        )

        /**
         * lastViewID: Stored last activated tab id,
         * hence convenient for disable the highlight icons.
         */
        private var lastViewID = R.id.home_tabs_menu_first


        /**
         * afterCreateHandler: It SHOULD be invoked AFTER the creation hook of activities.
         * 1. Register the homepage to be the first activated fragment.
         * 2. Set the OnClickListeners of every one of tabs through the context of current Fragment.
         */
        fun afterCreateHandler() {
            mapper[R.id.home_tabs_menu_first]?.let {
                getChildFragmentManager()
                    .beginTransaction()
                    .add(R.id.main_frame, it)
                    .commitNow()
            }
            lastViewID = R.id.home_tabs_menu_first

            val frame = (this.view as RelativeLayout?)!!

            frame.findViewById<HorizontalScrollView>(R.id.home_tabs_menu)
                .fullScroll(View.FOCUS_LEFT)

            mapper.forEach { (id: Int, _) ->
                val homeTab = frame.findViewById<TextView>(id)
                homeTab.setOnClickListener {
                    // If they equal, there's no need to change the icons.
                    if (id == lastViewID) return@setOnClickListener

                    // Clear highlights of the last activated tabs.
                    val lastTab = frame.findViewById<TextView>(lastViewID)
                    lastTab.setTypeface(null, Typeface.NORMAL)
                    ContextCompat.getColor(
                        frame.context,
                        R.color.home_tabs_original
                    ).let {
                        lastTab.setTextColor(it)
                    }

                    // Make highlights of new clicked tabs.
                    lastViewID = id
                    val thisTab = frame.findViewById<TextView>(id)
                    thisTab.setTypeface(null, Typeface.BOLD)
                    ContextCompat.getColor(
                        frame.context,
                        R.color.home_tabs_selected
                    ).let {
                        thisTab.setTextColor(it)
                    }

                    // Actually switch the layout.
                    mapper[id]?.let {
                        getChildFragmentManager()
                            .beginTransaction()
                            .replace(R.id.main_frame, it)
                            .commitNow()
                    }

                    // Auto Scroll
                    frame.findViewById<HorizontalScrollView>(R.id.home_tabs_menu)?.let {
                        when (id) {
                            mapper.keys.first() -> it.fullScroll(View.FOCUS_LEFT)
                            mapper.keys.last() -> it.fullScroll(View.FOCUS_RIGHT)
                        }
                    }
                }
            }
        }
    }

    class ChatPageFragment(id: Int) : AppTabFragment(id) {
        private var recyclerViewChat: RecyclerView? = null
        private var editTextMessage: EditText? = null
        private var buttonSend: Button? = null
        private var messageAdapter: MessageAdapter? = null
        private var messages: ArrayList<String>? = null

        fun afterCreateHandler() {
            val frame = (this.view as RelativeLayout?)!!
            recyclerViewChat = frame.findViewById(R.id.recycler_view_chat)
            editTextMessage = frame.findViewById(R.id.edit_text_message)
            buttonSend = frame.findViewById(R.id.button_send)

            messages = ArrayList()
            messageAdapter = MessageAdapter(messages!!)
            recyclerViewChat!!.setAdapter(messageAdapter)
            recyclerViewChat!!.setLayoutManager(LinearLayoutManager(this.context))

            buttonSend!!.setOnClickListener {
                val message = editTextMessage!!.getText().toString()
                if (message.isNotEmpty()) {
                    sendMessage(message)
                    sendMessage("You sent: $message", true)
                }
            }
        }

        private fun sendMessage(message: String, left: Boolean = false) {
            messages!!.add(message)
            messageAdapter!!.notifyItemInserted(messages!!.size - 1)
            recyclerViewChat!!.scrollToPosition(messages!!.size - 1)
            editTextMessage!!.getText().clear()
        }
    }

    /**
     * mapper: Store the home tabs and their respective DIY Fragments.
     */
    private val mapper = mapOf(
        R.id.navigation_home to HomePageFragment(R.layout.home_layout),
        R.id.navigation_chat to ChatPageFragment(R.layout.chat_layout),
        R.id.navigation_dashboard to AppTabFragment(R.layout.dashboard_layout)
    )

    /**
     * resMap: Store the home tabs and their respective dual icons
     * for activated and non-activated mode.
     */
    private val resMap = mapOf(
        R.id.navigation_home to (R.drawable.home_original to R.drawable.home_clicked),
        R.id.navigation_chat to (R.drawable.chat_original to R.drawable.chat_clicked),
        R.id.navigation_dashboard to (R.drawable.dashboard_original to R.drawable.dashboard_clicked)
    )

    /**
     * currentID: Store the last activated tab.
     */
    private var currentID = R.id.home_tabs_menu


    /**
     * handleNavigation: It follows the onCreate hook of main activity.
     * 1. Activate the home tab.
     * 2. Setup listeners.
     */
    private fun handleNavigation() {
        // Activate the home tab.
        activity.findViewById<TextView>(R.id.navigation_home)
            .setCompoundDrawablesWithIntrinsicBounds(
                null,
                ContextCompat.getDrawable(activity, R.drawable.home_clicked),
                null, null
            )
        currentID = R.id.navigation_home
        val defaultFragment =
            mapper[R.id.navigation_home] as HomePageFragment
        val supportFragmentManager = activity.supportFragmentManager
        supportFragmentManager
            .beginTransaction()
            .add(R.id.fragment_container, defaultFragment)
            .commitNow()

        mapper.forEach { (key: Int, value: AppTabFragment?) ->
            val bottomNavView = activity.findViewById<TextView>(key)
            bottomNavView.setOnClickListener {
                // If they equal, there's no need to change the icons.
                if (currentID == key) return@setOnClickListener

                // Clear the last activated flags.
                val last = activity.findViewById<TextView>(currentID)
                val viewOriginal = resMap[currentID]!!.first
                ContextCompat.getDrawable(activity, viewOriginal).let {
                    last.setCompoundDrawablesWithIntrinsicBounds(
                        null, it, null, null)
                }

                currentID = key

                // Activate current fragment.
                val here = activity.findViewById<TextView>(key)
                val viewClicked = resMap[key]!!.second
                val drawableClicked = ContextCompat.getDrawable(activity, viewClicked)
                here.setCompoundDrawablesWithIntrinsicBounds(null, drawableClicked, null, null)
                supportFragmentManager
                    .beginTransaction()
                    .replace(R.id.fragment_container, value as Fragment)
                    .commitNow()
                // Consequent callbacks
                when(key) {
                    R.id.navigation_home ->
                        (mapper[R.id.navigation_home] as HomePageFragment).let {
                            it.requireView()
                            it.afterCreateHandler()
                        }
                    R.id.navigation_chat ->
                        (mapper[R.id.navigation_chat] as ChatPageFragment).let {
                            it.requireView()
                            it.afterCreateHandler()
                        }
                }
            }
        }
    }

    /**
     * SHOULD be ONLY invoked after callbacks of the onCreate hook.
     */
    fun afterCreateHandle() {
        (mapper[R.id.navigation_home] as HomePageFragment).let {
            it.requireView()
            it.afterCreateHandler()
        }
    }

    /**
     * Can be invoked in onCreate hook of main activity.
     */
    fun onCreateHandle() {
        handleNavigation()
    }
}
